mesonbuild/scripts/reprotest.py

```diff
@@ -0,0 +1,128 @@
+# SPDX-License-Identifier: Apache-2.0
+# Copyright 2024 The Meson development team
+
+from __future__ import annotations
+
+import sys, os, subprocess, shutil
+import pathlib
+import typing as T
+
+if T.TYPE_CHECKING:
+    import argparse
+
+from ..mesonlib import get_meson_command
+
+# Note: when adding arguments, please also add them to the completion
+# scripts in $MESONSRC/data/shell-completions/
+def add_arguments(parser: 'argparse.ArgumentParser') -> None:
+    parser.add_argument('--intermediaries',
+                        default=False,
+                        action='store_true',
+                        help='Check intermediate files.')
+    parser.add_argument('mesonargs', nargs='*',
+                        help='Arguments to pass to "meson setup".')
+
+IGNORE_PATTERNS = ('.ninja_log',
+                   '.ninja_deps',
+                   'meson-private',
+                   'meson-logs',
+                   'meson-info',
+                   )
+
+INTERMEDIATE_EXTENSIONS = ('.gch',
+                           '.pch',
+                           '.o',
+                           '.obj',
+                           '.class',
+                           )
+
+class ReproTester:
+    def __init__(self, options: T.Any):
+        self.args = options.mesonargs
+        self.meson = get_meson_command()[:]
+        self.builddir = pathlib.Path('buildrepro')
+        self.storagedir = pathlib.Path('buildrepro.1st')
+        self.issues: T.List[str] = []
+        self.check_intermediaries = options.intermediaries
+
+    def run(self) -> int:
+        if not pathlib.Path('meson.build').is_file():
+            sys.exit('This command needs to be run at your project source root.')
+        self.check_ccache()
+        self.cleanup()
+        self.build()
+        self.check_output()
+        self.print_results()
+        if not self.issues:
+            self.cleanup()
+        return len(self.issues)
+
+    def check_ccache(self) -> None:
```

Comment:
I have `/usr/lib/ccache` in my `PATH` 
and that makes me currently not abort with `sys.exit(1)` even though I should.

```
> $ ccache -z 
Statistics zeroed

> $ time ~/git/meson/meson.py reprotest --intermediaries 
The Meson build system
Version: 1.5.99
Source dir: /home/fortysixandtwo/git/calls
Build dir: /home/fortysixandtwo/git/calls/buildrepro
Build type: native build

[...]

[422/422] Linking target tests/application
No differences detected.
~/git/meson/meson.py reprotest --intermediaries  94,72s user 21,80s system 491% cpu 23,705 total

> $ ccache -s
Cacheable calls:    550 / 633 (86.89%)
  Hits:             550 / 550 (100.0%)
    Direct:         548 / 550 (99.64%)
    Preprocessed:     2 / 550 ( 0.36%)
  Misses:             0 / 550 ( 0.00%)
Uncacheable calls:   83 / 633 (13.11%)
Local storage:
  Cache size (GiB): 1.3 / 5.0 (25.31%)
  Hits:             550 / 550 (100.0%)
  Misses:             0 / 550 ( 0.00%)

```

Reply:
Detecting all ways ccache could be injected is tricky, especially since actual compiler detection code is elsewhere and not easily callable. I guess we could check if `PATH` contains the string `ccache`, but that might have false positives.

Reply:
Simplest approach is to just use $CCACHE_DISABLE in the environment to tell ccache that even when run, it should simply forward directly to the compiler without performing any actions itself.

Reply:
This would also mean we could simply support the ccache users, as their testing would work by telling ccache to bypass the actual cache, instead of just refusing to perform the reproducibility testing at all.